### PR TITLE
Parallel docs

### PR DESCRIPTION
**Pull request by Sl1mb0**
This consolidates all the documentation I could find in the rustc-dev-guide regarding parallelism into one place; and describes parallelism at individual stages of the compiler.

 I plan to update this documentation with more details for parallelism at each stage of the compiler in subsequent PRs.

### CODE REVIEW

**Review by spastorino — APPROVED** [review]

File: src/parallel-rustc.md [hunk 1]
```diff
@@ -1,16 +1,55 @@
...
+parallel compiler work by enabling it in the `config.toml`.
+
+These next few sections describe where and how parallelism is currently used, 
+and the current status of making parallel compilation the default in `rustc`.
+
+The underlying thread-safe data-structures used in the parallel compiler 
+can be found in `rustc_data_structures/sync.rs`. Some of these data structures
+use the `parking_lot` API.
+
+## Code Gen
```
Comment (jyn514) [inline src/parallel-rustc.md:16]: I haven't seen it abbreviated this way before, it seems a little odd.
```suggestion
## Codegen
```

File: src/parallel-rustc.md [hunk 2]
```diff
@@ -1,16 +1,55 @@
...
+  we are done.
+- If there is no cache entry and no other active query invocation computing the
+  same result, we mark the key as being "in progress", release the lock and
+  start evaluating.
+- If there *is* another query invocation for the same key in progress, we
+  release the lock, and just block the thread until the other invocation has
+  computed the result we are waiting for. This cannot deadlock because, as
+  mentioned before, query invocations form a DAG. Some thread will always make
+  progress.
+
```
Comment (jyn514) [inline src/parallel-rustc.md:45]: Could you add a section on rustdoc as well? There's an overview here: https://github.com/rust-lang/rust/issues/82741

File: src/parallel-rustc.md (hunk 1 above)
Comment (Sl1mb0) [reply src/parallel-rustc.md:16]: To clarify:

Code Gen -> Codegen ?

Comment (jyn514) [reply src/parallel-rustc.md:16]: Yup, that's what I meant :)

File: src/parallel-rustc.md (hunk 2 above)
Comment (jyn514) [reply src/parallel-rustc.md:45]: Thanks!

**Review by jyn514 — APPROVED** [review]

**Pull request merged**